lib/treelify/acts_as_user.rb

```diff
@@ -14,22 +14,23 @@ def acts_as_user
     module InstanceMethods
       def role_for(resource)
         ansestors_ids = resource.hierarchy.self_and_ancestors_ids
-        Role.joins(:members)
-            .where("((roles.inherited = 't' "\
-                   "AND members.hierarchy_id IN (#{ansestors_ids.join(',')})) "\
-                   "OR (members.hierarchy_id = #{resource.hierarchy.id})) AND members.user_id = #{id}")
-            .order(level: :desc).first
+        Treelify::Role.joins(:members)
+                      .where("((treelify_roles.inherited = 't' "\
```

Comment:
I used your fixes version
```
where(treelify_roles: { inherited: true }, treelify_members: { hierarchy_id: ancestors_ids + [resource.hierarchy.id], user_id: id })
```
but tests were failing cuz of nested conditional with `OR` which isn`t supported yet in Active record.

But it will be available from Rails 5 :) http://juanitofatas.com/2015/01/31/rails-5-active-record-or/